taichi/analysis/clone.cpp

```diff
@@ -111,22 +110,25 @@ class IRCloner : public IRVisitor {
     }
   }
 
-  static std::unique_ptr<IRNode> run(IRNode *root) {
+  static std::unique_ptr<IRNode> run(IRNode *root, Kernel *kernel) {
+    if (kernel == nullptr) {
+      kernel = &get_current_program().get_current_kernel();
+    }
     std::unique_ptr<IRNode> new_root = root->clone();
     IRCloner cloner(new_root.get());
     cloner.phase = IRCloner::register_operand_map;
     root->accept(&cloner);
     cloner.phase = IRCloner::replace_operand;
     root->accept(&cloner);
-    irpass::typecheck(new_root.get());
+    irpass::typecheck(new_root.get(), kernel);
```

Comment:
Actually this pass is probably unnecessary...

Reply:
...since all `ret_type`s are cloned.

Reply:
Merging this in for now and let's get rid of all kernel parameters and this type check after #988 is done.